[IPsec] refactor xfrm policy state handling oss

ldelossa opened the pull request:
```release-note
Refactor XFRM policy and state creation
```

ldelossa committed:
ipsec: introduce new enableIPSec related functions

Introduce a new set of enableIPSec{4,6} related functions which split up
the duty of creating XFRM states and policies.

There are two logically separate code paths with regards to
creating/updating XFRM policies. One where "SubnetEncryption" is enabled
by IPAM and one where its disabled.

The introduced functions handle each case separately, reducing the
footprint of the original function which handles both cases.

A subsequent commit will cut over to using these new functions.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
ipsec: refactor enableIPsecIPv{4,6} to utilize helper methods

Update both the enableIPsecIPv4 and 6 methods to utilize the helper
methods introduced in the previous commit.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
ipsec: make fwd policy creation explicit

Prior to this commit the XFRM state and policy creation code would both
explicitly create XFRM FWD rules and implicitly (done during policy-in
creation).

This commit disambiguates where FWD policies are created. We opt to
always make them explicit. This will help with refactoring efforts in
a subsequent commit.

After this commit all FWD rules are created within the same function
that IN and OUT policies are created as well. We also add a new IPSecDir
constant to explicitly instruct the control plane to create a forward
rule.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
ipsec: introduce IPSecParameters structure

This commit introduces a new structure for declarative declaration of an
XFRM Policy and State.

This structure provides a consistent set of fields which are required
for creating an XFRM state and its associated policies.

Having a well defined structure allows us to declare exactly how an XFRM
policy should look at a higher level, and methods which are passed this
structure are in charge of creating the policy as the structure defines.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
ipsec: utilize declarative XFRM policy/state creation

This commit updates the methods which create XFRM states and policies to
use the declarative structure introduced in the previous commit.

Each policy is now well-defined before passing the declarative structure
down to the methods which do the heavy lifting of policy creation.

The heavy-lifting structures no longer impose a constraint on how/what
arguments are provided and instead refer to the declarative
IPSecParameter structure for the information they need.

This allows the developer to reason about the XFRM state and policy they
are creating at a higher level, and in one place.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
ipsec: remove TestUpsertIPSecEndpoint test

This function will be replaced by individual tests for IN, FWD, and OUT
policy creationg in the following commit.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
ipsec: update IPSec tests to use declarative XFRM creation

This commit introduces 3 new tests for specifically testing the creation
of XFRM states and policies in a particular direction, using the
declarative IPSecParamater struct.

Additionally, existing tests are updated to utilize the structure.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

jschwinger233 commented on pkg/datapath/linux/ipsec.go:
```diff
@@ -307,138 +307,36 @@ func (n *linuxNodeHandler) enableIPSecIPv4DoLocalHost(newNode *nodeTypes.Node, e
-			}
+	// If we are the local node, we have much less work to do, handle this first.
+	var isLocal bool
+	if isLocal, errs = n.enableIPSecIPv4DoLocalHost(newNode, errs); isLocal {
```
nit: how about moving `if newNode.IsLocal()` from inside `enableIPSecIPvXDoLocalHost()` to here:
```
if newNode.IsLocal() {
    n.enableIPSecIPvXDoLocalHost()
}
```

Because I feel it's more natural and consistent? The subnetEncryption branch is handled same way:

```
if n.subnetEncryption() {
    return n.enableIPSecIPv4DoSubnetEncryption(newNode, nodeID, zeroMark, updateExisting, errs)
}
```

ldelossa replied:
Updated.

jschwinger233 commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -580,17 +579,6 @@ func _ipSecReplacePolicyInFwd(src, dst *net.IPNet, tmplSrc, tmplDst net.IP, prox
-	// We always make forward rules optional. The only reason we have these
-	// at all is to appease the XFRM route hooks, we don't really care about
-	// policy because Cilium BPF programs do that.
-	if dir == netlink.XFRM_DIR_FWD {
```
Since dir==FWD is gone, the function name can be simplified to `_ipSecReplacePolicyIn` without a dir parameter?

ldelossa replied:
Yup, this is handled in a later commit, did you see this? https://github.com/cilium/cilium/pull/35210/commits/c41e71f14ee437b559e9026a8bda0823432f6f2a

jschwinger233 commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -539,22 +539,22 @@ func xfrmKeyEqual(s1, s2 *netlink.XfrmState) bool {
-func ipSecReplaceStateIn(log *slog.Logger, localIP, remoteIP net.IP, nodeID uint16, zeroMark bool, localBootID, remoteBootID string, remoteRebooted bool, reqID int) (uint8, error) {
-	key := getNodeIPsecKey(localIP, remoteIP, localBootID, remoteBootID, netlink.XFRM_DIR_IN)
+func ipSecReplaceStateIn(log *slog.Logger, params *IPSecParameters) (uint8, error) {
+	key := getNodeIPsecKey(*params.SourceTunnelIP, *params.DestTunnelIP, params.LocalBootID, params.RemoteBootID, netlink.XFRM_DIR_IN)
```
nit: this function looks suitable to become a method of *IPSecParameters

ldelossa replied:
I don't disagree, but this PR already makes a lot of changes. So may hold this one off until a follow-up.

jschwinger233 reviewed (commented):
High-level intent is clear, I left some nonblocking comments.

The commits in my opinion may not be split in a perfect way: the first commit introduces 3 functions representing 3 branches in enableIPsecIPvX(), but reviewer could have difficulty in confirming whether any logic has changed. Maybe a better way to organize these changes can be: introduce new functions commit by commit, remove the old branch replaced by new function in the same commit of new function.

New struct IPSecParameters looks good to me.

ldelossa committed:
ipsec: remove IPSecDirBoth

Remove the use of IPSecDirBoth for creating both IN and OUT policies.

This option was only used in tests and had no functional use during
Cilium's runtime.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

jschwinger233 reviewed (approved):
Thank you!

pchaigno commented on pkg/datapath/linux/ipsec.go:
```diff
@@ -304,138 +304,35 @@ func (n *linuxNodeHandler) enableIPSecIPv4DoLocalHost(errs error) (bool, error)
-	wildcardCIDR := &net.IPNet{IP: wildcardIP, Mask: net.IPv4Mask(0, 0, 0, 0)}
-
 	err := ipsec.IPsecDefaultDropPolicy(n.log, false)
+	if err != nil {
```
This condition is already implemented by `errors.Join()`: https://pkg.go.dev/errors#Join.

ldelossa replied:
Aha, the IPv6 branch had this, and I copied it from there thinking we may have missed there. I'll update both IPv4 and IPv6 functions to reflect this.

ldelossa replied:
Fixed with latest push

ldelossa replied:
Hmm, this change seemed to break test. Let me seeee

ldelossa replied:
Bleh, I see, the fix for this comment winded up creating an error **unconditionally** and adding it to the errs variable. Let me adjust..

ldelossa replied:
Fixed

pchaigno commented on pkg/datapath/linux/ipsec.go:
```diff
@@ -369,6 +373,10 @@ func (n *linuxNodeHandler) enableIPSecIPv6DoSubnetEncryption(newNode *nodeTypes.
 
+		if err = ipsec.IpSecReplacePolicyFwd(wildcardCIDR6, localIP, ipsec.DefaultReqID); err != nil {
+			n.log.Warn("egress unable to replace policy fwd", logfields.Error, err)
+		}
```
In commit `ipsec: make fwd policy creation explicit`, I see you're adding four `IpSecReplacePolicyFwd` statements in `pkg/datapath/linux/ipsec.go`. For `enableIPSecIPv{4,6}Do` and for `enableIPSecIPv6DoSubnetEncryption`, but not for `enableIPSecIPv4DoSubnetEncryption`. Makes sense because `enableIPSecIPv4DoSubnetEncryption` already had it. Does that mean we had a potential bug on the IPv6 codepath before due to the missing fwd policy?

ldelossa replied:
I don't think so.

Before this PR, FWD policies were created twice (per endpoint event) for IPv4 (I guess by error), and only once for IPv6 (probably the 'right' way).

In the IPv6 case, FWD rules were created, implicitly, by creating the IN policies. 

https://github.com/cilium/cilium/blob/main/pkg/datapath/linux/ipsec.go#L387 -> https://github.com/cilium/cilium/blob/main/pkg/datapath/linux/ipsec/ipsec_linux.go#L909

Now, we just make it explicit during each node event.


As discussed over slack, in a subsequent PR we'll reduce this to just a single FWD policy being created when IPsec is enabled.

pchaigno replied:
I see. That makes sense. Thanks!

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -99,6 +99,55 @@ type oldXfrmStateKey struct {
 }
 
+type IPSecParameters struct {
+	// The BootID for the local host, used for bookkeeping.
```
Not sure what you mean by "used for bookkeeping"?

ldelossa replied:
Yeah, I just swept that explanation under the rug since it doesn't deal directly with the resulting XFRM state or policy parameters. 

I'll update the wording on this a bit better. By bookkeeping I meant, more of the administrative task of whether or not to create a policy.

pchaigno replied:
It's also used for the key derivation.

ldelossa replied:
Fixed with latest push

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -99,6 +99,55 @@ type oldXfrmStateKey struct {
+	// decrypted the the packet.
+	DestTunnelIP *net.IP
+	// The ReqID used for the resulting XFRM policy/state
+	ReqID int
```
Could we make this higher-level? I think this one could be a boolean `EncryptedOverlay` with an appropriate description.

ldelossa replied:
IMO, these parameters should more or less plainly spell out how the resulting policy/state looks. I don't want to have hidden resulting  values in the policy which can't be seen from the high level structure. 

But even without this concern, in a future update, we will be removing the encrypted overlay distinction all together. It's kinda a hack which we have a solution for. Happy to talk more in detail about this in real time if you'd like.

pchaigno replied:
Sure. I only commented this because you wrote:
> Having a well defined structure allows us to declare exactly how an XFRM
policy should look at a higher level

But I don't mind either way :smiley_cat:

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -47,6 +47,7 @@ type IPSecDir string
 	IPSecDirOut     IPSecDir = "IPSEC_OUT"
+	IPSecDirFwd     IPSecDir = "IPSEC_FWD"
 	IPSecDirBoth    IPSecDir = "IPSEC_BOTH"
 	IPSecDirOutNode IPSecDir = "IPSEC_OUT_NODE"
```
FYI, https://github.com/cilium/cilium/pull/35346. Although that shouldn't be a concern as I'd expect your PR to be merged first.

ldelossa replied:
Nice! Want to push this forward after this merges? It looks good to me.

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -309,10 +309,7 @@ func getNodeIPsecKey(localNodeIP, remoteNodeIP net.IP, localBootID, remoteBootID
-	if dir == netlink.XFRM_DIR_OUT {
-		return deriveNodeIPsecKey(globalKey, localNodeIP, remoteNodeIP, localBootID, remoteBootID)
-	}
-	return deriveNodeIPsecKey(globalKey, remoteNodeIP, localNodeIP, remoteBootID, localBootID)
```
Commit `ipsec: remove direction from getNodeIPsecKey`:

Clearly, this is a functional change. But you're saying the PR doesn't change the behavior here (or tests wouldn't pass anyway). So I assume that means a previous commit included the inverse change in the caller. AFAIU, that means the previous commit includes a silent, breaking change. Could we avoid it and ensure no commit breaks IPsec (or if they do, at least call it out explicitly)?

ldelossa replied:
Yup that's fair, I caught this while testing. I can collapse this into into a relevant previous commit 👍

ldelossa replied:
Fixed with latest push

pchaigno replied:
Nop, nop :slightly_smiling_face: 

You still needed to invert the boot IDs :wink:

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -632,10 +632,10 @@ func _ipSecReplacePolicyInFwd(params *IPSecParameters, proxyMark bool, dir netli
+	if err := _ipSecReplacePolicyIn(params, true, netlink.XFRM_DIR_IN); err != nil {
 		return err
 	}
-	return _ipSecReplacePolicyInFwd(params, false, netlink.XFRM_DIR_IN)
```
Commit `ipsec: use appropriate name for policy in creation function`:

I'm all for small commits, but I'm going to have to agree with Gray on this one: this is just bugging reviewers when we review the previous commit and notice the function wasn't renamed then. Plus, extracting those three lines doesn't exactly simplify the previous commit.

ldelossa replied:
This one was not done for the same reason as splitting the first two commits. This, i just saw later on after I finished the bulk of this change. I'll collapse this. 👍

ldelossa replied:
Fixed with latest push

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux_test.go:
```diff
@@ -15,6 +15,7 @@ import (
 	"github.com/stretchr/testify/require"
```
On commit `ipsec: update IPSec tests to use declarative XFRM creation`:

Since we're on the topic of the perfect commit history:
> Additionally, existing tests are updated to utilize the structure.

This first word, in a commit description, is usually a good sign that the commit could be split :stuck_out_tongue_winking_eye:

pchaigno reviewed (commented):
Mostly nits, two small concerns below:
- I'm wondering if one commit isn't actually a bugfix for IPv6.
- One commit seems to break the key derivation, then another fixes it, all silently :grimacing:
